components/num-util/src/fixed_decimal.rs

```diff
@@ -0,0 +1,517 @@
+    /// dec.is_negative = true;
+    /// assert_eq!("-0", dec.to_string());
+    /// ```
+    pub is_negative: bool,
+}
+
+impl Default for FixedDecimal {
+    /// Returns a FixedDecimal representing zero.
+    fn default() -> Self {
+        FixedDecimal {
+            digits: SmallVec::new(),
+            magnitude: 0,
+            upper_magnitude: 0,
+            lower_magnitude: 0,
+            is_negative: false,
+        }
+    }
+}
+
+impl From<isize> for FixedDecimal {
+    fn from(value: isize) -> Self {
+        let value_abs = if value == std::isize::MIN {
+            // isize::MIN.abs() is too big for isize by 1
+            std::isize::MAX as usize + 1
+        } else {
+            value.abs() as usize
+        };
+        let mut result = FixedDecimal::from_ascending(UintIterator::from(value_abs)).unwrap();
+        result.is_negative = value < 0;
+        result
+    }
+}
+
+/// A wrapper over an unsigned integer, used when constructing a FixedDecimal.
+/// This is a workaround for: https://github.com/rust-lang/rust/issues/50133
+pub struct UnsignedWrapper<T: Unsigned + Integer + FromPrimitive + ToPrimitive>(pub T);
+
+impl<T> TryFrom<UnsignedWrapper<T>> for FixedDecimal
+where
+    T: Unsigned + Integer + FromPrimitive + ToPrimitive,
+{
+    type Error = Error;
+
+    /// Create a FixedDecimal from an unsigned integer.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use icu_num_util::FixedDecimal;
+    /// use icu_num_util::fixed_decimal::UnsignedWrapper;
+    /// use std::convert::TryInto;
+    ///
+    /// let dec: FixedDecimal = UnsignedWrapper(750usize).try_into().unwrap();
+    /// assert_eq!("750", dec.to_string());
+    /// ```
+    fn try_from(value: UnsignedWrapper<T>) -> Result<Self, Self::Error> {
+        FixedDecimal::from_ascending(UintIterator::from(value.0))
+    }
+}
+
+impl FixedDecimal {
+    /// Initialize a FixedDecimal with an iterator of digits in ascending
+    /// order of magnitude, starting with the digit at magnitude 0.
+    ///
+    /// This method is not public; use TryFrom::<UintIterator<T>> instead.
+    fn from_ascending<T>(digits_iter: T) -> Result<Self, Error>
+    where
+        T: Iterator<Item = u8>,
+    {
+        // TODO: make X a usize generic to customize the size of this array
+        // https://github.com/rust-lang/rust/issues/44580
+        const X: usize = 32;
+        // A temporary structure to allow the digits in the iterator to be reversed.
+        // The digits are inserted started from the end, and then a slice is copied
+        // into its final destination (result.digits).
+        let mut mem: [u8; X] = [0u8; X];
+        let mut trailing_zeros: usize = 0;
+        let mut i: usize = 0;
+        for (x, d) in digits_iter.enumerate() {
+            // Take only up to std::i16::MAX values so that we have enough capacity
+            if x > std::i16::MAX as usize {
+                return Err(Error::Limit);
+            }
```

Comment:
`digits_iter.take(std::i16::MAX).enumerate()`

Reply:
and then you can check if the iterator is done at the end and throw if its not.

You can also use `ExactSizeIterator` which is reasonable I think in this case - I don't expect anyone to throw streaming lexer at this and try to parse digit by digit an unknown length source.

Reply:
> and then you can check if the iterator is done at the end and throw if its not.

How do you do that?  `.take()` moves the iterator so I can't call any methods on it later, right?

> You can also use ExactSizeIterator which is reasonable I think in this case - I don't expect anyone to throw streaming lexer at this and try to parse digit by digit an unknown length source.

Actually no, because the common case is that you're pulling digits off an integer.  To determine how many digits are in the integer, you would need to either perform an expensive `log10` on the input, or run the number through a series of if/else statements checking the number against different powers of 10.  The benefit of an ExactSizeIterator doesn't seem to outweigh the cost.

Reply:
> How do you do that? .take() moves the iterator so I can't call any methods on it later, right?

uh, right! Sorry for noise. I was confused about the role of `take`.